Add SEMIANNUALLY option to Utility Meter

williamobrien opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Adding in a BIANNUALLY cycle for utility cycles that cover a 6 month period.


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml
utility_meter:
  energy:
    source: sensor.energy_in_kwh
    cycle: biannually
```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [ ] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

williamobrien committed:
Adding a BIANNUALLY option for 6 month cycles

williamobrien committed:
Update const.py (#1), merging it with the updated sensor.py

dgomes commented:
Please add a test

williamobrien commented:
> Please add a test

Hello - I'm new to contributing to Home Assistant, what test do I need to perform?

Thank you!

williamobrien commented:
> > Please add a test
> 
> Hello - I'm new to contributing to Home Assistant, what test do I need to perform?
> 
> Thank you!

Oh, wait - I looked back at when the bimonthly option was added and might have figured out what you mean, I'll see if I can fix it!

frenck commented:
Let's make this configurable in other ways. We can keep adding time frame likes this, it just feels counterproductive.

williamobrien committed:
Patch 2 (#2)

* Update zha to use new fan entity model (#45758)

* Update zha to use new fan entity model

* fixes

* tweaks for zha

* pylint

* augment test cover

* Centralize wemo exception handling (#46705)

* Perform wemo state update quickly after a timeout (#46702)

* Perform state update quickly after a timeout

* with async_timeout.timeout(...) -> async with async_timeout.timeout(...)

* [ci skip] Translation update

* Bump pylutron 0.2.7 (#46717)

* Switch ssdp to be async by using async_upnp_client for scanning (#46554)

SSDP scans no longer runs in the executor

This is an interim step that converts the async_upnp_client
response to netdisco's object to ensure fully backwards
compatibility

* Add selectors to Input * service definitions (#46652)

* Add selectors to Alert service definitions (#46627)

* Add selectors to Camera service definitions (#46630)

* Add selectors to Switch service definitions (#46635)

* Add selectors to Downloader service definitions (#46638)

* Add selectors to Fan service definitions (#46639)

* Add selectors to Cover service definitions (#46634)

Co-authored-by: Tobias Sauerwein <cgtobi@users.noreply.github.com>

* Add selectors to Automation service definitions (#46629)

* Add selectors to Climate service definitions (#46632)

* Add selectors to Alarm Control Panel service definitions (#46626)

* Upgrade and constrain httplib2>=0.19.0 (#46725)

* ubus: switch to pypi library (#46690)

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

* Add selectors to Counter service definitions (#46633)

* Add advanced service parameter flag (#46727)

* Add discontinuity support to HLS streams and fix nest expiring stream urls (#46683)

* Support HLS stream discontinuity.

* Clarify discontinuity comments

* Signal a stream discontinuity on restart due to stream error

* Apply suggestions from code review

Co-authored-by: uvjustin <46082645+uvjustin@users.noreply.github.com>

* Simplify stream discontinuity logic

* Add selectors to Scene service definitions (#46729)

* Fix missing color switch specific device class for Z-Wave JS driver >6.3 (#46718)

* Add selectors to Netatmo services (#46574)

* Add selectors

* Fix schedul selector

* Update homeassistant/components/netatmo/services.yaml

Co-authored-by: Bram Kragten <mail@bramkragten.nl>

* Update homeassistant/components/netatmo/services.yaml

Co-authored-by: Bram Kragten <mail@bramkragten.nl>

* Update services.yaml

* Added required field

Co-authored-by: Bram Kragten <mail@bramkragten.nl>

* Add test for biannually cycle in utility_meter component

Co-authored-by: J. Nick Koston <nick@koston.org>
Co-authored-by: Eric Severance <esev@esev.com>
Co-authored-by: HomeAssistant Azure <hello@home-assistant.io>
Co-authored-by: uvjustin <46082645+uvjustin@users.noreply.github.com>
Co-authored-by: Franck Nijhof <git@frenck.dev>
Co-authored-by: Tobias Sauerwein <cgtobi@users.noreply.github.com>
Co-authored-by: Álvaro Fernández Rojas <noltari@gmail.com>
Co-authored-by: Allen Porter <allen@thebends.org>
Co-authored-by: Jesse Campbell <hikeit@gmail.com>
Co-authored-by: Bram Kragten <mail@bramkragten.nl>

williamobrien commented:
OK, so I added tests in for the "biannually" option, does that work?

dgomes commented:
you need to rebase and force push.

williamobrien committed:
Revert "Patch 2 (#2)"

This reverts commit e542852deb050185d1393f214b44d71ad3a6943b.

williamobrien committed:
Add tests for biannually meter timeframe

williamobrien commented:
> you need to rebase and force push.

OK, i tried to do that - so hopefully, it worked

dgomes commented:
You need to run lint the test, check the CI errors

williamobrien committed:
Fixed formatting and CI Errors

dgomes commented:
I have a big concern: https://dictionary.cambridge.org/dictionary/english/biannually

biannually means different things :(

williamobrien commented:
> I have a big concern: https://dictionary.cambridge.org/dictionary/english/biannually
> 
> biannually means different things :(

Yeah, I know - before I started making this, I struggled with trying to pick the best word to use, as there isn't really a great word for "every half year" (like something like "quarter"). I think in general the options were between biannually, biennially and semiannually.

Like you said - the problem is that Biannually could mean twice a year or once every 2 years. I think Biennial always means every 2 years, so that doesn't work. Maybe semiannually (https://dictionary.cambridge.org/dictionary/english/semi-annual?q=semiannual) is a better option?

williamobrien committed:
Changed cycle from BIANNUALLY to SEMIANNUALLY for clarity

johnsdixon commented:
Having just added half-hourly, I concur that the cron option is probably best and most flexible for the longer term. Code (c1a836fb9184dbb0592917d5d4653734984b86f8) and test(0f7e2ae171c2f7d5612753388d7217b42fcc76fd) for the half-hourly period if needed.

> Note to self: Check the issues for other similar problems to the one you're trying to fix, prior to embarking on a journey with a deprecated end-point.

dgomes commented:
@williamobrien and @johnsdixon could you give the cron PR a test ? 

I think we should close this PR and future ones and favour that route

johnsdixon commented:
Will do, I'll add comments to the cron PR.

johnsdixon commented:
@williamobrien
`croniter` (the parser/iterator in the alternative PR) supports standard `cron` entries. This would allow `cron: "0 0 1 */6 *` for twice yearly resets (January, July) or `cron: "0 0 1 4,10 *"` for resets in April and October.  `cron: "0 0 1 4/6 *"` is also supported, but is not as readable to my mind. HTH.

williamobrien commented:
@dgomes and @johnsdixon - I agree that it makes much more sense to use the fully customizable cron entries found in #46795 rather than continually creating new, one-off time periods like in my PR. I'll give it a go testing the alternative PR (although I'm still working on getting a dev environment setup for HA) and comment on it in the new PR!
